core/domain/base_model_validators.py

```diff
@@ -84,12 +88,57 @@ def __init__(
             model_class: ClassObject. The external model class.
             model_ids: list(str). The list of external model ids to fetch the
                 external models.
+
+        Raises:
+            Exception. This class was used instead of using the
+                UserSettingsModelFetcherDetails for the UserSettingsModel.
         """
+        if model_class == user_models.UserSettingsModel:
+            raise Exception(
+                'Please use the UserSettingsModelFetcherDetails ' +
+                'for UserSettingsModel')
         self.field_name = field_name
         self.model_class = model_class
         self.model_ids = model_ids
 
 
+class UserSettingsModelFetcherDetails(python_utils.OBJECT):
+    """Value object providing the class and ids to fetch the
+    user settings model.
+    """
+
+    def __init__(
+            self, field_name, model_ids,
+            system_user_ids_removed=False, pseudonymous_ids_removed=False):
```

Comment:
Require these two to always be specified; have no defaults. The caller should always be thinking about and specifying this.

Also, in the docstring for these args, please explain clearly when it is appropriate to specify True vs False for each of them, so that devs who are adding new validators know how to make the decision.

Reply:
Actually, before you make more changes, can you give a reply to the second paragraph above -- i.e. what are the actual decision criteria? That will help clarify what the defaults should be and what measures should be taken to prevent any mistakes from happening.

Reply:
@seanlip, The models that should allow these ids only after they contain these ids or find the necessary. The decision for which models should allow system user ids and pseudo user ids are suggested by @vojtechjelinek , So I guess he can give a better reasoning to this. Please also refer to this [discussion](https://github.com/oppia/oppia/pull/11245#discussion_r543881208) regarding default values.

Reply:
Thanks -- I've read the discussion but I think my opinion still stands. I also need to understand what the rules are for setting these booleans before proceeding with the review. Could you please work with @vojtechjelinek on figuring out the explicit guidelines?

Reply:
I would say those explicit guidelines are:

The field should allow pseudonymized IDs when the model should be pseudonymized when the wipeout is requested.

The field should allow admin IDs when some kind of admin or bot can create/modify these kinds of models as part of some CRON or a one-off job. We should always try to use a user ID when possible, but in some cases, like schema migrations or automatic generation of suggestions, there is no explicit user actor.

Reply:
Hm. Some follow-up questions:

- Maybe we should have the bools reversed instead, e.g. pseudonymous_ids_should_be_included (if not including them is what is typically expected)?
- Couldn't the pseudonymous ids status be inferred from the methods already defined on the class for Wipeout?

Also, should we be handling the case where we don't expect "special user ID" references to exist but there are some? I.e. there should be some sort of error logging if, say, we don't expect admin user IDs but those exist in the list of models to fetch.

Reply:
> Maybe we should have the bools reversed instead, e.g. pseudonymous_ids_should_be_included (if not including them is what is typically expected)?

Yep, I think that naming would also be okay, I think there is the slight problem that when we create this model from outside the IDs are indeed included, but inside the constructor, we actually remove them from the list of IDs to analyze and then the name `*_be_inclued` might be a bit confusing.

> Couldn't the pseudonymous ids status be inferred from the methods already defined on the class for Wipeout?

I think it would be better to keep it in the validators for now (as some models allow pseudo only in some fields and not in others). In long term, the links between models should be may be implemented on the model definition level, with the introduction of custom field type like `UserIdField` that will indicate that this field can only contain user IDs and should correspond to `UserSettingsModel` (or be pseudo/admin). This is definitely planned for next year.

> Also, should we be handling the case where we don't expect "special user ID" references to exist but there are some? I.e. there should be some sort of error logging if, say, we don't expect admin user IDs but those exist in the list of models to fetch.

This is done by default since the fetch for UserSettingsModels will fail for these IDs, but it might be good to skip these by default.

Reply:
Ah ok, thanks. One question though -- for the last point: 

> Also, should we be handling the case where we don't expect "special user ID" references to exist but there are some? I.e. there should be some sort of error logging if, say, we don't expect admin user IDs but those exist in the list of models to fetch.

>> This is done by default since the fetch for UserSettingsModels will fail for these IDs, but it might be good to skip these by default.

Ah, I see. Yup, I think it's better to do it explicitly since it's conceptually different.

I.e., let's say model X has a field F which contains user IDs. There are four cases depending on whether F is allowed to contain system IDs, pseudonymous IDs, or both. 

The following errors might occur:
- Entries in F are invalid (the definition of "invalid" depends on which case F falls into)
- Entries in F do not correspond to a user settings model when we try to do the retrieval

For the first case I propose that the fetcher method should catch these and report errors. Currently I think it just does set subtraction.
For the second case I propose that we report any "missing model" errors.

Does this make sense / do you agree?

---

Also, going back to the proposed guidelines:

> The field should allow pseudonymized IDs when the model should be pseudonymized when the wipeout is requested.
>
> The field should allow admin IDs when some kind of admin or bot can create/modify these kinds of models as part of some CRON or a one-off job. We should always try to use a user ID when possible, but in some cases, like schema migrations or automatic generation of suggestions, there is no explicit user actor.

Let's write these more "directive-ly", as instructions to developers:

> Set may_contain_spurious_pseudonymized_ids to True if and only if this field can contain user IDs that are pseudonymized as part of Wipeout.
> Set may_contain_spurious_system_ids to True if and only if this field can contain admin or bot IDs.

Are these interpretations correct? Also I have suggested new names for these args in the description above based on my current understanding, what do you think of them?

Reply:
> Does this make sense / do you agree?

Yes.

> Are these interpretations correct? Also I have suggested new names for these args in the description above based on my current understanding, what do you think of them?

Yes. Maybe we can add a note to the pseudo that these fields can only be in models that have `LOCALLY_PSEUDONYMIZE` as their `DELETION_POLICY`. The naming seems okay, I would maybe use just `may_contain_pseudonymized_ids` and `may_contain_system_ids`, because I don't see the added value of adding "spurious" there, to be honest, the usage of "spurious" confuses me here.

Reply:
Sounds good, I agree, let's drop "spurious".

In addition to adding a note, how about checking this condition within the fetcher and throwing an error if it is not satisfied?

Reply:
> In addition to adding a note, how about checking this condition within the fetcher and throwing an error if it is not satisfied?

Reply:
OK, sgtm. @rohitkatlaa do you have enough information to proceed?

If so, please feel free to assign me once you're ready for a full review. Thanks!

Reply:
@vojtechjelinek @seanlip 
My tasks:
- Rename system_user_ids_omitted to may_contain_system_ids
- Write a note to developers regarding when these bools should be set to True.
- Check if the DELETION_POLICY as LOCALLY_PSEUDONYMIZE then check if may_contain_pseudonymized_ids is True else throw error.
Did I cover all the tasks?

And regarding the task 3: In the fetcherDetails class there are only field name, ids(of userSettingModel) information, I do not know which validator is requesting, so I won't be able to check the deletion policy. So how should I implement it?

Reply:
There's more: avoid these args having default values, and also see my comment [here](https://github.com/oppia/oppia/pull/11245#discussion_r549523606) about general error-reporting.

I'll defer to @vojtechjelinek to advise on the third point.

Reply:
@rohitkatlaa Good point. I cannot think of any simple straightforward way to implement it. The only idea that I have is to have a test in `AuditJobsTestBase` that takes the output of `_get_external_id_relationships` and verify that the classes do not have `may_contain_pseudonymized_ids` set to True when the model does not have the `DELETION_POLICY` set to `LOCALLY_PSEUDONYMIZE`.

Reply:
> There's more: avoid these args having default values, and also see my comment here about general error-reporting.

@seanlip Regarding this, 

> Entries in F do not correspond to a user settings model when we try to do the retrieval

The missing model error already occurs when the audit jobs are run.

> Entries in F are invalid (the definition of "invalid" depends on which case F falls into), 
> > For the first case I propose that the fetcher method should catch these and report errors. Currently I think it just does set subtraction.

I am a bit confused as to what invalid means here.(Assuming, invalid means invalid pseudo id, in that case I am removing only valid pseudo ids so invalid pseudo ids will cause an model missing error)

Reply:
@vojtechjelinek 
> I cannot think of any simple straightforward way to implement it. The only idea that I have is to have a test in AuditJobsTestBase that takes the output of _get_external_id_relationships and verify that the classes do not have may_contain_pseudonymized_ids set to True when the model does not have the DELETION_POLICY set to LOCALLY_PSEUDONYMIZE

Even in AuditJobsTestBase, there is no direct way to access the validator class, it has job_class from which I have to fetch the validator name and then import it and then check for the condition

Reply:
> I am a bit confused as to what invalid means here.(Assuming, invalid means invalid pseudo id, in that case I am removing only valid pseudo ids so invalid pseudo ids will cause an model missing error)

@rohitkatlaa Ah, sorry, let me clarify. What I mean is this: let's say that, in the fetcher, may_contain_admin_ids is False. Then if an admin ID is discovered in the list of IDs to fetch, we should error. Does that clarify? (Similarly for pseudonymous IDs.)